usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2695,6 +2703,282 @@ static CK_RV ccatok_pkey_add_attrs(STDLL_TokData_t * tokdata, TEMPLATE *template
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Create the protected key by iterating over all APQNs */
+    memset(&pkey_wrap_handler_data, 0, sizeof(pkey_wrap_handler_data_t));
+    pkey_wrap_handler_data.tokdata = tokdata;
+    pkey_wrap_handler_data.secure_key = skey_attr->pValue;
+    pkey_wrap_handler_data.secure_key_len = skey_attr->ulValueLen;
+    pkey_wrap_handler_data.pkey_buf = (CK_BYTE *)&pkey_buf;
+    pkey_wrap_handler_data.pkey_buflen_p = &pkey_buflen;
+    pkey_wrap_handler_data.keytype = key_type;
+
+    ret = cca_iterate_adapters(tokdata, ccatok_pkey_sec2prot,
+                               &pkey_wrap_handler_data);
+
+    if (ret != CKR_OK || !pkey_wrap_handler_data.wrap_was_successful) {
+        TRACE_ERROR("cca_iterate_adapters failed or no APQN could create the pkey.\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Build new attribute for protected key */
+    ret = build_attribute(CKA_IBM_OPAQUE_PKEY, pkey_buf, pkey_buflen, &tmp_attr);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("build_attribute failed with rc=0x%lx\n", ret);
+        ret = CKR_FUNCTION_FAILED;;
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+
+    *pkey_attr = tmp_attr;
+
+    return ret;
+}
+
+/*
+ * Save the current firmware wrapping key verification pattern in the tokdata:
+ * create a dummy test key, transform it into a protected key, and store the wk
+ * verification pattern in the tokdata.
+ */
+static CK_RV ccatok_pkey_get_firmware_wkvp(STDLL_TokData_t *tokdata)
+{
+    struct cca_private_data *cca_data = tokdata->private_data;
+    CK_ATTRIBUTE *pkey_attr = NULL, *sec_attr = NULL;
+    CK_RV ret;
+    long return_code = 0, reason_code = 0;
+    long exit_data_len = 0, clear_key_bit_length = 0;
+    unsigned char exit_data[4];
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE];
+    long rule_array_count;
+    unsigned char *clear_key_value = NULL;
+    long key_name_length = 0;
+    unsigned char key_name[CCA_KEY_ID_SIZE] = { 0, };
+    long user_data_length = 0;
+    unsigned char  user_data[64] = { 0, };
+    long token_data_length = 0;
+    long verb_data_length = 0;
+    unsigned char verb_data[64] = { 0, };
+    unsigned char the_key_token[CCA_KEY_TOKEN_SIZE] = { 0, };
+    long the_key_token_length = sizeof(the_key_token);
+    unsigned char key_type_1[8] = { ' ', };
+    unsigned char key_type_2[8] = { ' ', };
+    long key_name_1_length = 0;
+    long key_name_2_length = 0;
+    long user_data_1_length = 0;
+    long user_data_2_length = 0;
+    long kek_identifier_1_length = 0;
+    long kek_identifier_2_length = 0;
+    long the_2nd_key_token_length = 0;
+
+    /* Build CPACF-exportable AES internal cipher key token */
+    rule_array_count = 5;
+    memcpy(rule_array, "INTERNAL" "NO-KEY  " "AES     " "CIPHER  " "XPRTCPAC",
+           rule_array_count * CCA_KEYWORD_SIZE);
+
+    dll_CSNBKTB2(&return_code, &reason_code, &exit_data_len, exit_data,
```

Comment:
So I guess this will fail if it is sent to a crypto adapter that does not support the XPRTCPAC keyword, or if the host library does not support it. So this is an implicit check for the require versions.

However: What if multiple CCA adapters are in the system, and DEV-ANY is used, and not all of the adapters are on the desired level? 

Also You create an AES key with XPRTCPAC, this is actually supported with CEX6C already. But EC with XPRTCPAC needs CEX7C I guess..... So this check here will actually check for CEX6C, but not for CEX7C.

I guess you will need to add an explicit version check of all APQNs used by this token. Iterate over them using cca_iterate_adapters() and check their FW version/level. 

Only if ALL adapters are at least CEX6C, then XPRTCPAC can be used for AES keys.
Only if ALL adapters are at least CEX7C, then XPRTCPAC can be used for ECC keys.

So you might want to differentiate between AES and EC key support of protected key, or yo just enable PKEY only if all are CEX7Cs. 

You must make sure to not attempt to create a key with the XPRTCPAC  keyword, if not all adapters selectable for this token supports it.

Reply:
I would actually also add an explicit check for the CCA holt library version to be 7.x.x. or later. You can use field `version` in struct cca_private_data, it contains version, release, modification level of the used host library already.

For AES keys, version 6.3.27 would be sufficient, but I guess we can just require 7.0.0 or later.

Reply:
Please also check the adapter versions as mentioned in my first comment on this place.

Reply:
> Please also check the adapter versions as mentioned in my first comment on this place

See new commit: add check for minimum card version